File: next-app/src/lib/tomtomClient.ts
'use server';

function getTomTomKey() {
  const key = process.env.TOMTOM_API_KEY;
  if (!key) {
    throw new Error('TOMTOM_API_KEY is not configured. Please check your .env.local file.');
  }
  return key;
}

/**
 * Maps ORS profiles to TomTom travel modes
 */
function mapProfileToTomTom(profile: string): string {
  switch (profile) {
    case 'driving-car': return 'car';
    case 'foot-walking': return 'pedestrian';
    case 'cycling-regular': return 'bicycle';
    default: return 'car';
  }
}

/**
 * Fetches a duration matrix (in minutes) from TomTom with live traffic.
 * Fits within the 100-element limit for synchronous requests (e.g., 10x10 matrix).
 */
export async function getTomTomDurationsMatrix(coords: [number, number][], profile: string = 'driving-car'): Promise<number[][] | null> {
  const apiKey = getTomTomKey();
  
  if (coords.length <= 1) return [[0]];
  if (coords.length > 20) {
    console.warn('TomTom Matrix limit exceeded (>20 stops). Falling back to base durations.');
    return null;
  }

  // TomTom expects { point: { latitude, longitude } }
  const points = coords.map(([lat, lon]) => ({
    point: { latitude: lat, longitude: lon }
  }));

  const url = `https://api.tomtom.com/routing/1/matrix/json?key=${apiKey}&traffic=true&travelMode=${mapProfileToTomTom(profile)}`;

  try {
    const res = await fetch(url, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        origins: points,
        destinations: points
      }),
      next: { revalidate: 300 } // Balance live traffic: 5 minute cache
    });

    if (!res.ok) {
      console.error('TomTom Matrix API failed:', await res.text());
      return null;
    }

    const data = await res.json();
    
    // Matrix v1 structure: data.matrix[originIndex][destinationIndex]
    // Each entry has a "response" or "statusCode": 200
    if (!data.matrix) return null;

    const matrix: number[][] = data.matrix.map((row: any) => 
      row.map((cell: any) => {
        const durationSeconds = cell.response?.routeSummary?.travelTimeInSeconds;
        return durationSeconds !== undefined ? durationSeconds / 60 : 99999;
      })
    );

    return matrix;
  } catch (err) {
    console.error('TomTom Client Error:', err);
    return null;
  }
}

/**
 * Fetches detailed travel stats for a single leg, including historical usuals.
 */
export async function getTomTomLegDetails(origin: [number, number], destination: [number, number], profile: string = 'driving-car') {
  const apiKey = getTomTomKey();
  const mode = mapProfileToTomTom(profile);
  
  // traffic=all + departAt=now triggers the return of live vs historical
  const url = `https://api.tomtom.com/routing/1/calculateRoute/${origin[0]},${origin[1]}:${destination[0]},${destination[1]}/json?key=${apiKey}&traffic=true&travelMode=${mode}&departAt=now`;

  try {
    const res = await fetch(url, { next: { revalidate: 300 } });
    if (!res.ok) return null;
    const data = await res.json();
    const route = data.routes?.[0]?.summary;
    if (!route) return null;

    const live = route.travelTimeInSeconds / 60;
    const historical = route.historicTrafficTravelTimeInSeconds / 60;
    const noTraffic = route.noTrafficTravelTimeInSeconds / 60;

    return {
      liveMinutes: live,
      historicalMinutes: historical || live || noTraffic || 0.1,
      noTrafficMinutes: noTraffic || live || 0.1
    };
  } catch (err) {
    return null;
  }
}

File: api/clients/tomtom_client.py
import requests
from typing import List, Tuple, Optional, Dict
from ..config import settings

def map_to_tomtom_mode(profile: str) -> str:
    """Maps ORS profiles to TomTom travel modes"""
    mapping = {
        "driving-car": "car",
        "foot-walking": "pedestrian",
        "cycling-regular": "bicycle"
    }
    return mapping.get(profile, "car")

from ..engine.cache_manager import traffic_cache, get_cached_item, set_cached_item
import json

def get_tomtom_durations_matrix(coords: List[Tuple[float, float]], profile: str = "car", traffic: bool = True) -> Optional[List[List[float]]]:
    if not settings.TOMTOM_API_KEY:
        return None

    # V8.5: Cache Check
    coords_key = "|".join([f"{lat:.4f},{lon:.4f}" for lat, lon in coords])
    cache_key = f"matrix:{coords_key}:mode:{profile}:traffic:{traffic}"
    
    cached = get_cached_item(traffic_cache, cache_key)
    if cached:
        return cached

    # TomTom expects [lat, lon]
    origins = [{"point": {"latitude": lat, "longitude": lon}} for lat, lon in coords]
    destinations = origins
    
    traffic_param = "true" if traffic else "false"
    mode = map_to_tomtom_mode(profile)
    url = f"https://api.tomtom.com/routing/1/matrix/sync/json?key={settings.TOMTOM_API_KEY}&routeType=fastest&traffic={traffic_param}&travelMode={mode}"
    
    try:
        res = requests.post(
            url,
            json={
                "origins": origins,
                "destinations": destinations
            },
            timeout=10
        )

        if not res.ok:
            print(f"DEBUG: TomTom Matrix API Error: {res.status_code} {res.text}")
            return None
        
        data = res.json()
        matrix = []
        n = len(coords)
 
        # Verify matrix structure
        if "matrix" not in data or len(data["matrix"]) < n:
            return None
 
        for i in range(n):
            row = []
            for j in range(n):
                cell = data["matrix"][i][j]
                # TomTom returns travelTimeInSeconds
                seconds = cell.get("response", {}).get("routeSummary", {}).get("travelTimeInSeconds")
                # Fallback to high value if path not found
                row.append(seconds / 60 if seconds is not None else 99999.0)
            matrix.append(row)
        
        # Cache store
        set_cached_item(traffic_cache, cache_key, matrix)
        return matrix
    except Exception as e:
        print(f"DEBUG: TomTom Matrix Exception: {e}")
        return None

def get_tomtom_route_summary(coords: List[Tuple[float, float]], profile: str = "car") -> List[Dict]:
    """
    Fetches detailed travel stats for an entire fixed sequence of coordinates.
    Used to bypass the 100-cell matrix limit for the final schedule generation.
    Supports up to 50 waypoints.
    """
    if not settings.TOMTOM_API_KEY or len(coords) < 2:
        return []

    # V8.5: Cache Check
    coords_key = "|".join([f"{lat:.4f},{lon:.4f}" for lat, lon in coords])
    cache_key = f"summary:{coords_key}:mode:{profile}"
    
    cached = get_cached_item(traffic_cache, cache_key)
    if cached:
        return cached

    # TomTom expects {lat},{lon}:{lat},{lon}...
    points_str = ":".join([f"{lat},{lon}" for lat, lon in coords])
    mode = map_to_tomtom_mode(profile)
    url = f"https://api.tomtom.com/routing/1/calculateRoute/{points_str}/json?key={settings.TOMTOM_API_KEY}&traffic=true&travelMode={mode}&departAt=now&computeTravelTimeFor=all"
    
    try:
        res = requests.get(url, timeout=15)
        if not res.ok:
            print(f"DEBUG: TomTom Route Summary Error: {res.status_code} {res.text}")
            return []
        
        data = res.json()
        route = data.get("routes", [{}])[0]
        legs = route.get("legs", [])
        
        results = []
        for leg in legs:
            summary = leg.get("summary", {})
            results.append({
                "liveMinutes": summary.get("travelTimeInSeconds", 0) / 60,
                "historicalMinutes": summary.get("historicTrafficTravelTimeInSeconds", summary.get("travelTimeInSeconds", 0)) / 60,
                "noTrafficMinutes": summary.get("noTrafficTravelTimeInSeconds", summary.get("travelTimeInSeconds", 0)) / 60
            })
        
        # Cache store
        if results:
            set_cached_item(traffic_cache, cache_key, results)
        return results
    except Exception as e:
        print(f"DEBUG: TomTom Route Summary Exception: {e}")
        return []

def get_tomtom_leg_details(start: Tuple[float, float], end: Tuple[float, float], profile: str = "car") -> Optional[Dict]:
    """Legacy individual leg fetch (fallback)"""
    res = get_tomtom_route_summary([start, end], profile)
    return res[0] if res else None
